src/plugin/debugger/mhpdbgc.c

```diff
@@ -345,86 +355,134 @@ static int check_for_stopped(void)
+
+    num++;
   }
+  if (user_symbol_num < num)
+    user_symbol_num = num;
+
+  fclose(fp);
+  return num;
+}
+
+static void usermap_clear(void)
+{
+  memset(&user_symbol, 0, sizeof user_symbol);
+  user_symbol_num = 0;
+}
 
-  if (!getval_ul(argv[1], 16, &org)) {
-    mhp_printf("origin parse error '%s'\n", argv[1]);
+static void mhp_usermap(int argc, char *argv[])
+{
+  unsigned int origin;
+  int num;
+
+  if (argc < 2 ||
+      (strcmp(argv[1], "list") != 0 &&
+       strcmp(argv[1], "load") != 0 &&
+       strcmp(argv[1], "clear") != 0)) {
+    mhp_printf("syntax: usermap load <file> [origin]\n");
+    mhp_printf("syntax: usermap clear\n");
+    mhp_printf("syntax: usermap list\n");
     return;
   }
-  symbl2_org = org;
 
-  ifp = fopen(argv[2], "r");
-  if (!ifp) {
-     mhp_printf("unable to open map file %s\n", argv[2]);
-     return;
-  }
+  if (strcmp(argv[1], "list") == 0) {
+    int i;
 
-  mhp_printf("reading map file %s\n", argv[2]);
-  last_symbol2 = 0;
-  for(;;) {
-     if(!fgets(bytebuf, sizeof (bytebuf), ifp)) {
-        mhp_printf("error: could not find following in %s:\n%s\n",
-                   argv[2], srchfor);
-        return;
-     }
-     if (!strlen(bytebuf))
-        continue;
-     if (!memcmp(bytebuf, srchfor, strlen(srchfor)))
-        break;
+    for (i = 0; i < user_symbol_num; i++) {
+      if (user_symbol[i].name[0])
+        mhp_printf("  %04x:%04x %s %s\n",
+                   user_symbol[i].seg, user_symbol[i].off,
+                   user_symbol[i].type == ABS ? "ABS" : "   ",
+                   user_symbol[i].name);
+    }
+    return;
   }
-  for(;;) {
-     if(!fgets(bytebuf, sizeof (bytebuf), ifp))
-        break;
-     if(bytebuf[5] != ':')
-        continue;
-     if(bytebuf[12] != ' ')
-        continue;
-     bytebuf[5] = ' ';
-     sscanf(&bytebuf[1], "%x %x", &seg, &off);
-     symbl2_table[last_symbol2].seg  = seg + (symbl2_org >> 4);
-     symbl2_table[last_symbol2].off  = off;
-     symbl2_table[last_symbol2].type = ' ';
-     sscanf(&bytebuf[17], "%s", (char *)&symbl2_table[last_symbol2].name);
-     last_symbol2++;
+
+  if (strcmp(argv[1], "clear") == 0) {
+    usermap_clear();
+    return;
   }
-  fclose(ifp);
-  if (!last_symbol2) {
-    mhp_printf("failed to read symbols from map file\n");
+
+  // load
+  if (argc < 3) {
+    mhp_printf("error: load requires a map file argument\n");
     return;
   }
-/*symbl2_end = symbl2_table[last_symbol2-1].addr;*/
-  mhp_printf("%d symbol(s) processed\n", last_symbol2);
-  mhp_printf("highest address %04x:%04x(%s)\n",
-             symbl2_table[last_symbol2-1].seg,
-             symbl2_table[last_symbol2-1].off,
-             symbl2_table[last_symbol2-1].name);
 
+  if (argc >= 4) {
+    if (!getval_ui(argv[3], 16, &origin)) {
+      mhp_printf("error: origin parse error '%s'\n", argv[3]);
+      return;
+    }
+  } else {
+    origin = 0;
+  }
+
+  num = usermap_load_file_mslink(argv[2], origin);
```

Comment:
It would be better to just write errors inside
the func itself, and only return err/ok to the
caller.
Btw, why "mslink"? How about gnu ld?

Reply:
`usermap_load_file_mslink` is the existing map file parsing just split out to a separate function.

I had figured on creating a `usermap_load_file_gnuld` and returning the same error codes, but I certainly can put them back in the parse function.

Reply:
If you want a common error handling
for printing, just create a separate function
for that. There are many attempts to
standardize the way of returning the optional
error info to user (`errno` in posix, exceptions
in C++ etc), but using the funky enums for that
is almost always an anti-pattern.

Reply:
okay, done.